src/gsMultiGrid/gsGridHierarchy.h

```diff
@@ -0,0 +1,373 @@
+template <typename T>
+void coarsenMultiBasis_withTransfer(
+    const gsMultiBasis<T>& mBasis,
+    const gsBoundaryConditions<T>& boundaryConditions,
+    const gsOptionList& assemblerOptions,
+    gsMultiBasis<T> & coarsenedMBasis,
+    gsSparseMatrix<T, RowMajor>& transferMatrix,
+    std::vector< gsSparseMatrix<T, RowMajor> >& localTransferMatrices
+    );
+
+/// @brief This function coarsens the bases of a gsMultiBasis and provides the transfer matrix.
+///
+/// @param mBasis[in]                          The gsMultiBasis to be refined
+/// @param boundaryConditions[in]              The boundary conditions
+/// @param assemblerOptions[in]                A gsOptionList defining a "DirichletStrategy" and a "InterfaceStrategy"
+/// @param coarsenedMBasis[out]                A pointer to the coarsened gsMultiBasis
+/// @param transferMatrix[out]                 The transfer matrix for the free dofs
+///
+/// \ingroup Solver
+template <typename T>
+inline void coarsenMultiBasis_withTransfer(
+    const gsMultiBasis<T>& mBasis,
+    const gsBoundaryConditions<T>& boundaryConditions,
+    const gsOptionList& assemblerOptions,
+    gsMultiBasis<T> & coarsenedMBasis,
+    gsSparseMatrix<T, RowMajor>& transferMatrix
+    )
+{
+    std::vector< gsSparseMatrix<T, RowMajor> > localTransferMatrices;
+    coarsenMultiBasis_withTransfer( mBasis, boundaryConditions, assemblerOptions, coarsenedMBasis, transferMatrix, localTransferMatrices );
+}
+
+/// @brief This function takes local transfer matrices (per patch) and combines them
+/// using given DofMappers to a global transfer matrix. Simultanously, this
+/// function restricts the matrices to the free dofs, e.g., Dirichlet dofs might be
+/// eliminated
+///
+/// @param localTransferMatrices[in]     The local and full (also non-free dofs) transfer matrices per patch
+/// @param coarseMapper[in]              The DofMapper on the coarse grid
+/// @param fineMapper[in]                The DofMapper on the fine grid
+/// @param transferMatrix[out]           The combined transfer matrix restricted to the free dofs
+///
+/// \ingroup Solver
+template <typename T>
+void combineTransferMatrices(
+    const std::vector< gsSparseMatrix<T, RowMajor> >& localTransferMatrices,
+    const gsDofMapper& coarseMapper,
+    const gsDofMapper& fineMapper,
+    gsSparseMatrix<T, RowMajor>& transferMatrix
+    );
+
+
+/// @brief Copy items from \a data into \a result unless their index is in \a indices. Order is preserved.
+///
+/// If an item's index is contained in \a indices, then it is dropped, otherwise it is copied into \a result.
+///
+/// @param data[in]                    Container to copy from
+/// @param indices[in]                 Vector of indices. The vector MUST be sorted in increasing order.
+/// @param result[out]                 Container to copy to
+///
+/// \ingroup Solver
+template <typename Cont>
+void copyIfNotIndexed(const Cont& data, const std::vector<index_t>& indices, Cont& result);
```

Comment:
this seams to duplicate http://en.cppreference.com/w/cpp/algorithm/set_difference